pkg/compact/downsample/downsample_test.go

```diff
@@ -266,6 +267,40 @@ func TestCounterSeriesIterator(t *testing.T) {
 	testutil.Equals(t, exp, res)
 }
 
+func TestCounterSeriesIteratorSeek(t *testing.T) {
+	defer leaktest.CheckTimeout(t, 10*time.Second)()
+
+	chunks := [][]sample{
+		{{100, 10}, {200, 20}, {300, 10}, {400, 20}, {400, 5}},
+	}
+	exp := []sample{
+		{100, 10}, {200, 20}, {300, 30}, {400, 40},
+	}
+
+	var its []chunkenc.Iterator
+	for _, c := range chunks {
+		its = append(its, newSampleIterator(c))
+	}
+
+	x := NewCounterSeriesIterator(its...)
+	var it *storage.BufferedSeriesIterator
```

Comment:
Here to simulate the function `matrixIterSlice` in Promql at [L1086](https://github.com/prometheus/prometheus/blob/master/promql/engine.go#L1086)

Reply:
Can we make this test actually bit shorter? Basically this test hides what is exactly tested within some `storage.BufferedSeriesIterator`. And it kinds of make sense with your comment about `matrixIterSlice` but it just makes this unit test, test more than required.

I think the right pattern here is just a test that `Seek` works, so:
- it gives false if there is no sample after or in this timestamp
- it gives true if opposite case it happening
- (regression case) it gives true when there is sample after seek timestamp, even when there is only one sample.

Third case will actually fail without you fix in `CounterSeriesIterator` and it is all that matter here without complicating unit test (: 

What do you think?

Reply:
Additionally, I would add just comment that the regression could be seen while using `matrixIterSlice` in `PromQL` - that might be useful.

Reply:
Thanks for your review and suggestions. 

At first I did not import the `storage.BufferedSeriesIterator`, but I think it will not reach the key point and trigger this bug. Because `storage.BufferedSeriesIterator` will  do something dependent on its internal `SeriesIterator` is `CounterSeriesIterator` actually. I thought it did not make sense to  implement `storage.BufferedSeriesIterator`, I use it directly.

It seems that I should force on the `CounterSeriesIterator` itself. And just mock the action `storage.BufferedSeriesIterator` does.

Reply:
> It seems that I should force on the CounterSeriesIterator itself. And just mock the action storage.BufferedSeriesIterator does.

You could do it. But what action it is doing, it is basically having an iterator that has only first sample and asking for Seek(0) on this (or something like this). So IMO these 3 test cases I proposed are totally enough. Do you agree? (:

Reply:
I agree with you. I will update the test unit.